sway-lsp/src/core/token.rs

```diff
@@ -162,22 +163,60 @@ impl Token {
         }
     }
 
-    /// Return the [Ident] of the declaration of the provided token.
-    pub fn declared_token_ident(&self, engines: &Engines) -> Option<Ident> {
+    /// Return the [TokenIdent] of the declaration of the provided token.
+    pub fn declared_token_ident(&self, engines: &Engines) -> Option<TokenIdent> {
         self.type_def.as_ref().and_then(|type_def| match type_def {
             TypeDefinition::TypeId(type_id) => ident_of_type_id(engines, type_id),
-            TypeDefinition::Ident(ident) => Some(ident.clone()),
+            TypeDefinition::Ident(ident) => Some(TokenIdent::new(ident, engines.se())),
         })
     }
+}
 
-    /// Return the [Span] of the declaration of the provided token. This is useful for
-    /// performaing == comparisons on spans. We need to do this instead of comparing
-    /// the [Ident] because the [PartialEq] implementation is only comparing the name.
-    pub fn declared_token_span(&self, engines: &Engines) -> Option<Span> {
-        self.type_def.as_ref().and_then(|type_def| match type_def {
-            TypeDefinition::TypeId(type_id) => Some(ident_of_type_id(engines, type_id)?.span()),
-            TypeDefinition::Ident(ident) => Some(ident.span()),
-        })
+/// A more convenient [Ident] type for use in the language server.
+///
+/// This type is used as the key in the [TokenMap]. It's constructed during AST traversal
+/// where we compute the [Range] of the token and the convert [SourceId]'s to [PathBuf]'s.
+/// Although this introduces a small amount of overhead while traversing, precomputing this
+/// greatly speeds up performace in all other areas of the language server.
+///
+/// [TokenMap]: crate::core::token_map::TokenMap
+/// [SourceId]: sway_types::SourceId
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
+pub struct TokenIdent {
+    pub name: String,
+    pub range: Range,
+    pub path: Option<PathBuf>,
+    pub is_raw_ident: bool,
+}
+
+impl TokenIdent {
+    pub fn new(ident: &Ident, se: &SourceEngine) -> Self {
+        let path = ident
+            .span()
+            .source_id()
+            .map(|source_id| se.get_path(source_id));
```

Comment:
Consider throwing an error here

Reply:
Ideally `source_id()` wouldn't return an `Option` as outlined above. This should be an error caught in a CI test so upstream tools consuming the AST's always have definite paths to work with.